src/ConfigUpdater.php

```diff
@@ -0,0 +1,279 @@
+        $id_key = $definition->getKey('id');
+
+        $id = $value[$id_key];
+        /** @var \Drupal\Core\Config\Entity\ConfigEntityStorageInterface $entity_storage */
+        $entity_storage = $this->entityTypeManager->getStorage($type);
+        /** @var \Drupal\Core\Config\Entity\ConfigEntityInterface $entity */
+        $entity = $entity_storage->load($id);
+        $uuid = $entity->get('uuid');
+        $entity = $entity_storage->updateFromStorageRecord($entity, $value);
+        $entity->set('uuid', $uuid);
+        $entity->save();
+      }
+
+      $succeeded[$config_name] = $this->t('@config was successfully reverted', ['@config' => $config_name]);
+    }
+
+    return new ConfigUpdaterResult($failed, $succeeded);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function delete(array $config_names) {
+    $succeeded = [];
+    $failed = [];
+
+    foreach ($config_names as $config_name) {
+      $value = $this->loadFromActive($config_name);
+      if (!$value) {
+        $failed[$config_name] = $this->t('@config did not exist in extension storage', ['@config' => $config_name]);
+        continue;
+      }
+
+      // Check the configuration object's hash and see if it has been modified.
+      if ($this->isModified($config_name)) {
+        $failed[$config_name] = $this->t('@config has been modified and was not deleted', ['@config' => $config_name]);
+        continue;
+      }
+
+      $type = $this->getConfigType($config_name);
+      if ($type == 'system.simple') {
+        $config = $this->configFactory->getEditable($config_name);
+        if (!$config) {
+          $failed[$config_name] = $this->t('@config did not exist in active storage', ['@config' => $config_name]);
+          continue;
+        }
+        $config->delete();
+      }
+      else {
+        $definition = $this->entityTypeManager->getDefinition($type);
+        $id_key = $definition->getKey('id');
+        $id = $value[$id_key];
+
+        /** @var \Drupal\Core\Config\Entity\ConfigEntityStorageInterface $entity_storage */
+        $entity_storage = $this->entityTypeManager->getStorage($type);
+        $entity = $entity_storage->load($id);
+        $entity_storage->delete([$entity]);
+      }
+
+      $succeeded[$config_name] = $this->t('@config was successfully deleted', ['@config' => $config_name]);
+    }
+
+    return new ConfigUpdaterResult($failed, $succeeded);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function loadFromActive($config_name) {
+    return $this->activeConfigStorage->read($config_name);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function loadFromExtension($config_name) {
+    $value = $this->extensionConfigStorage->read($config_name);
+    if (!$value) {
+      $value = $this->extensionOptionalConfigStorage->read($config_name);
+    }
+    return $value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function isModified($config_name) {
+    // Read the configuration from active storage.
```

Comment:
"Read the configuration from active storage." says the same thing as the line of code it's describing. Therefore it's redundant. We need to document the WHY and the non-obvious things. The comments below about the hashes are good examples of that (documenting non-obvious things).